gordon_janitor_gcp/cloud_dns.py

```diff
@@ -0,0 +1,125 @@
+
+
+@attr.s
+class GCPResourceRecordSet:
+    """DNS Resource Record Set.
+
+    Args:
+        name (str): Name/label.
+        type (str): Record type (see `Google's supported records
+            <https://cloud.google.com/dns/overview#supported_dns_r
+            ecord_types>`_ for valid types).
+        rrdatas (list): Record data according to RFC 1034§3.6.1 and
+            RFC 1035§5.
+        ttl (int): (optional) Number of seconds that the record set can
+            be cached by resolvers. Defaults to 300.
+    """
+    # TODO (lynn): This will be moved to a common package to be shared
+    #   between all of gordon* packages. It will also make use of attrs
+    #   ability to optionally validate upon creation.
+    name     = attr.ib(type=str)
+    type     = attr.ib(type=str)
+    rrdatas  = attr.ib(type=list)
+    ttl      = attr.ib(type=int, default=300)
+
+
+class AIOGoogleDNSClient(http_client.AIOGoogleHTTPClient):
+    """Async HTTP client to interact with Google Cloud DNS API.
+
+    Attributes:
+        BASE_URL (str): base call url for the DNS API
+
+    Args:
+        project (str): Google project ID that hosts the managed DNS.
+        keyfile (str): path to service account (SA) keyfile.
+        scopes (list): scopes with which to authorize the SA. Default is
+            ``['cloud-platform']``.
+        api_version (str): DNS API endpoint version. Defaults to ``v1``.
+        loop: asyncio event loop to use for HTTP requests.
+    """
+    BASE_URL = 'https://www.googleapis.com/dns'
+
+    def __init__(self, project=None, keyfile=None, scopes=None,
+                 api_version='v1', loop=None):
+        super().__init__(keyfile=keyfile, scopes=scopes, loop=loop)
+        self.project = project
+        self._base_url = f'{self.BASE_URL}/{api_version}/projects/{project}'
+
+    def _parse_records(self, response, records):
+        unparsed_records = response.get('rrsets', [])
+        for record in unparsed_records:
+            rrset = GCPResourceRecordSet(**record)
+            records.append(rrset)
+        return records
+
+    async def get_records_for_zone(self, zone):
+        """Get all resource record sets for a particular managed zone.
+
+        Args:
+            zone (str): Desired managed zone to query.
+        Returns:
+            list of :py:class:`GCPResourceRecordSet` instances.
+        """
+        url = f'{self._base_url}/managedZones/{zone}/rrsets'
+
+        # to limit the amount of data across the wire; also makes it
+        # easier to create GCPResourceRecordSet instances
+        fields = 'rrsets/name,rrsets/rrdatas,rrsets/type,rrsets/ttl'
+        params = {
+            'fields': fields,
+        }
+        next_page_token = None
+
+        records = []
+        while True:
+            if next_page_token:
+                params['pageToken'] = next_page_token
+            response = await self.get_json(url, params=params)
+            records = self._parse_records(response, records)
```

Comment:
This is a nitpick. I found it a bit confusing that the code here passes in `records`, the `_parse_records` method mutates the object and returns it, and the `records` variable gets reassigned to the same object. I think this combines two approaches that I've always seen used separately:

```python
# don't have to assign anything since the method modifies the list:
self._parse_records(response, records)
```

and

```python
# the caller takes care of modifying the list, the method just parses the input
records.extend(self._parse_records(response)
```

Do you think it'd be better to go with one of them? Or I could be missing something here.

Reply:
I'm not sure what the first approach is that much clearer if the idea is that it modifies the list behind the scenes. 

I often chose not to do a `list.extend` because it can turn into a noticeable performance hit. So I chose to just append within the called function.

¯\\_(ツ)_/¯

Reply:
I guess my confusion comes from `records` getting modified in two places on the same line - the caller and the callee. If I called `_parse_records` again like `records2 = []; records2 = self.parse_records(response, records)` then `id(records) == id(records2)` and records would contain a duplicate of each item. Maybe the method name could hint at this eg `_parse_rsp_into_records`?